### PR TITLE
#2584 adds support for OIDC login.

### LINKED ISSUE #2548: Keycloak Integration
**Is your feature request related to a problem? Please describe.**
NGLP are using Keycloak as their SSO solution.

**Describe the solution you'd like**
To make Janeway work nicely with other NGLP software (AD, WDP etc) we should add the ability to use Keycloak for authentication. Authorisation will remain part of Janeway.

**Describe alternatives you've considered**
N/a

**Additional context**
NGLP requires this.

### PR DESCRIPTION
**Pull request by ajrbyers**
- Adds support for OIDC thru Mozilla's Django OIDC package.
- Closes #2548

### CODE REVIEW

File: docs/source/oidc.rst [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+        'utils.oidc.JanewayOIDCAB',
+    )
+
+.. note:: All of the urls above are examples derived from a Keycloak instance. You can get the real URLs by using your service providers .well-known url which will look something like:     https://exmaple/org/.well-known/openid-configuration. You can get your Client ID and Secret from your service provider admin console or can request one from your technical staff. If you use RS256 for OIDC_RP_SIGN_ALGO you must also complete OIDC_OP_JWKS_ENDPOINT otherwise login will fail.
+
+New Accounts
+------------
+When a user without an account on Janeway logs in a new active account is generated for them with their Email address as the django username, first name and last name are also copied.
+
+If you want to stop OIDC from creating accounts you can add `OIDC_CREATE_USER = False` to your settings file.
```
Comment (mauromsl) [inline docs/source/oidc.rst:43]: Just curious as to what would happen in Janeway if this is set to False and I log in with an account that does not exist in Janeway

File: src/utils/oidc.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+        return user
+
+    def update_user(self, user, claims):
+        user.first_name = claims.get('given_name', '')
+        user.last_name = claims.get('family_name', '')
+        user.is_active = True
+        user.save()
+
+        return user
```
Comment (mauromsl) [inline src/utils/oidc.py:39]: Can we have a test case for these?

**Review by mauromsl — CHANGES_REQUESTED** [review]

File: src/utils/oidc.py (hunk 2 above)
Comment (ajrbyers) [reply src/utils/oidc.py:39]: No you can’t. Go away.

File: docs/source/oidc.rst (hunk 1 above)
Comment (ajrbyers) [reply docs/source/oidc.rst:43]: Dropped this line.

**Review by mauromsl — APPROVED** [review]
exciting!

**Pull request merged**